lib/srv/alpnproxy/local_proxy.go

```diff
@@ -0,0 +1,288 @@
+		return trace.Wrap(err)
+	}
+
+	if err = sess.RequestSubsystem(proxySubsystemName(l.cfg.SSHUserHost)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := proxySession(l.context, sess); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func proxySubsystemName(userHost string) string {
+	return fmt.Sprintf("proxy:%s", userHost)
+}
+
+func makeSSHClient(conn *tls.Conn, addr string, cfg *ssh.ClientConfig) (*ssh.Client, error) {
+	cc, chs, reqs, err := ssh.NewClientConn(conn, addr, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return ssh.NewClient(cc, chs, reqs), nil
+}
+
+func proxySession(ctx context.Context, sess *ssh.Session) error {
+	stdout, err := sess.StdoutPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stdin, err := sess.StdinPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stderr, err := sess.StderrPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	errC := make(chan error)
+	go func() {
+		_, err := io.Copy(os.Stdout, stdout)
+		errC <- err
+	}()
+	go func() {
+		_, err := io.Copy(stdin, os.Stdin)
+		errC <- err
+	}()
+	go func() {
+		_, err := io.Copy(os.Stderr, stderr)
+		errC <- err
+	}()
+	var errs []error
+	for i := 0; i < 3; i++ {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errC:
+			if err != nil && !errors.Is(err, io.EOF) {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// Start starts the LocalProxy.
+func (l *LocalProxy) Start(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		conn, err := l.cfg.Listener.Accept()
+		if err != nil {
+			if utils.IsOKNetworkError(err) {
+				return nil
+			}
+			log.WithError(err).Errorf("Faield to accept client connection.")
+			continue
+		}
+		go func() {
+			if err := l.handleDownstreamConnection(ctx, conn, l.cfg.SNI); err != nil {
+				log.WithError(err).Errorf("Failed to handle connection.")
+			}
+		}()
+	}
+}
+
+// GetAddr returns the LocalProxy listener address.
+func (l *LocalProxy) GetAddr() string {
+	return l.cfg.Listener.Addr().String()
+}
+
+func (l *LocalProxy) handleDownstreamConnection(ctx context.Context, downstreamConn net.Conn, serverName string) error {
+	defer downstreamConn.Close()
+	upstreamConn, err := tls.Dial("tcp", l.cfg.RemoteProxyAddr, &tls.Config{
+		NextProtos:         []string{string(l.cfg.Protocol)},
+		InsecureSkipVerify: l.cfg.InsecureSkipVerify,
+		ServerName:         serverName,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer upstreamConn.Close()
+
+	errC := make(chan error)
+	go func() {
+		defer downstreamConn.Close()
+		defer upstreamConn.Close()
+		_, err := io.Copy(downstreamConn, upstreamConn)
+		errC <- err
+	}()
+	go func() {
+		defer downstreamConn.Close()
+		defer upstreamConn.Close()
+		_, err := io.Copy(upstreamConn, downstreamConn)
+		errC <- err
+	}()
+
+	var errs []error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
+			return nil
```

Comment:
I'm not 100% sure this is valid syntax but maybe something like this would be good
```suggestion
			return trace.NewAggregate(ctx.Err(), errs...)
```